### ISSUE TITLE
numbered toc and sections without dot after the last digit

### ISSUE THREAD

**Issue by p627** [asciidoctor/asciidoctor-pdf#858]
Hi, I wonder if there is an opportunity to delete the dot in the numbering after the last digit? I think this is the common convention.

I have:

 
![grafik](https://user-images.githubusercontent.com/32871353/31664517-0bfd6644-b346-11e7-9d95-e4e306cce8cc.png)

I search for something like this:

 
![grafik](https://user-images.githubusercontent.com/32871353/31664523-124e9a9a-b346-11e7-8f24-ea74a6ca0ed4.png)

Thank you in advance.

**Comment by lisa-rosenberg** [comment]
I have the same problem.

janosvitok, can you tell me how to use your snippet, please? The command you have given is leading me to this output:

```
rosenbeli@N0601:/mnt/d/Documents$ asciidoctor-pdf -r remove-section-trailing-dot.rb
Usage: asciidoctor [OPTION]... FILE...
Translate the AsciiDoc source FILE or FILE(s) into the backend output format (e.g., HTML 5, DocBook 4.5, etc.)
By default, the output is written to a file with the basename of the source file and the appropriate extension.
Example: asciidoctor -b html5 source.asciidoc

    -b, --backend BACKEND            set output format backend: [html5, xhtml5, docbook5, docbook45, manpage] (default: html5)
                                     additional backends are supported via extensions (e.g., pdf, latex)
    -d, --doctype DOCTYPE            document type to use when converting document: [article, book, manpage, inline] (default: article)

[...]
```

Am I missing something? I am in the working directory where the .adoc file and the .rb file are stored.

I have to admit that I'm new to bash, asciidoctor and ruby. Maybe that's the problem :)

**Comment by janosvitok** [comment]
Lisa: Did you specify all your usual parameters in addition to the -r ? E.g. your document?
What I actually meant was that you have to add `-r remove-section-trailing-dot.rb` to your current asciidoctor parameters.

**Comment by lisa-rosenberg** [comment]
Hi janosvitok,

I couldn't get it to work until now, but I will look further into this, thank you :)

I am testing it with a very basic document:
```
= This is a Test

:numbered:
== With a section
=== A subsection
And some text.
```

After:
`asciidoctor-pdf -r remove-section-trailing-dot.rb -d book test.adoc --trace`

I'm getting this:
```
/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- remove-section-trailing-dot.rb (LoadError)
        from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from /usr/lib/ruby/vendor_ruby/polyglot.rb:63:in `require'
        from /var/lib/gems/2.3.0/gems/asciidoctor-1.5.6.1/lib/asciidoctor/cli/options.rb:230:in `block in parse!'
        from /var/lib/gems/2.3.0/gems/asciidoctor-1.5.6.1/lib/asciidoctor/cli/options.rb:228:in `each'
        from /var/lib/gems/2.3.0/gems/asciidoctor-1.5.6.1/lib/asciidoctor/cli/options.rb:228:in `parse!'
        from /var/lib/gems/2.3.0/gems/asciidoctor-pdf-1.5.0.alpha.16/bin/asciidoctor-pdf:25:in `<top (required)>'
        from /usr/local/bin/asciidoctor-pdf:22:in `load'
        from /usr/local/bin/asciidoctor-pdf:22:in `<main>'
```

Packages should be up to date:
```
ruby -v
ruby 2.3.5p376 (2017-09-14 revision 59905) [x86_64-linux-gnu]

asciidoctor -v
Asciidoctor 1.5.6.1 [http://asciidoctor.org]
Runtime Environment (ruby 2.3.5p376 (2017-09-14 revision 59905) [x86_64-linux-gnu]) (lc:UTF-8 fs:UTF-8 in:- ex:UTF-8)

asciidoctor-pdf -v
Asciidoctor PDF 1.5.0.alpha.16 using Asciidoctor 1.5.6.1 [http://asciidoctor.org]
Runtime Environment (ruby 2.3.5p376 (2017-09-14 revision 59905) [x86_64-linux-gnu]) (lc:UTF-8 fs:UTF-8 in:- ex:UTF-8)
```

**Comment by janosvitok** [comment]
@lisa-rosenberg 
https://stackoverflow.com/questions/9750610/ruby-require-error-cannot-load-such-file
http://asciidoctor.org/docs/user-manual/#cli-options

TL;DR: either provide path to remove-section-trailing-dot.rb (e.g. -r ./remove-section-trailing-dot.rb or -r /home/asdf/remove-section-trailing-dot.rb) or add -I path (e.g. -I . or -I /home/asdf)

**Comment by michael-o** [comment]
I'd be interested too. There is no trailing dot in numbering. This is against common convention.

**Comment by mojavelinux** [comment]
To move forward, we need a way to express this configuration. It probably belongs in the theme. We need to agree on a key and value. Implementation is then rather trivial I would suspect.

**Comment by mojavelinux** [comment]
(it could also be done with a document attribute so it also applies to other converters).

**Comment by mojavelinux** [comment]
I'm looking for the name of an option that may exist in other systems, so we can build on established conventions. If we can't find one, we'll have to invent one.

**Comment by dalemartin** [comment]
@mojavelinux what ever became of this? I would like to remove the trailing periods if it's been implemented.

**Comment by mojavelinux** [comment]
We're waiting on architecture. That hasn't been solved, so no code can be written yet. Feel free to propose solutions.

**Comment by michael-o** [comment]
I'd like  to see this removed by default. Haven't found this in any book type setting nor in LaTeX.

**Comment by mojavelinux** [comment]
We can't simply change behavior on all users. There has to be a setting. Some way to describe that you want different output than what is typical (for Asciidoctor).

**Comment by mojavelinux** [comment]
O'Reilly always includes the trailing dot in their books. Other publishers do not.

**Comment by dalemartin** [comment]
> We're waiting on architecture. 

Not sure what this means? But either way, not a huge deal to me, just nice to have at this point.

**Comment by mojavelinux** [comment]
In order to code it, I need to know what to code. Since there are arguments both for and against it, I've made the decision that it must be controlled through configuration (with the current behavior being the default). And that brings us to this comment:

> To move forward, we need a way to express this configuration.

Until we decide on what that configuration is, the specifications, nothing is going to happen. Feel free to submit proposals.

**Comment by janosvitok** [comment]
I'd say that add two doc attributes:
- :sectnumseparator:, by default '.'
- :sectnumappendseparator:, by default yes
and both asciidoctor and asciidoctor-pdf should be changed (and maybe more)

**Comment by mojavelinux** [comment]
Interesting proposal. How strongly do you feel about the fact that this should be controlled from the document? It seems to me something that is presentation only, and thus should be configured in the theme. (On the other hand, the HTML converter doesn't really have a theme, so we lose the opportunity to configure it there).

If we use the attributes, I would make it sectnumseparator and sectnumtrailingseparator (which defaults to sectnumseparator). That way, we could achieve `1-1.` or `1-1` or `1.1.` or `1.1`.

**Comment by janosvitok** [comment]
I don't have any strong feelings. In fact, I was thinking about something similar, but I've forgotten it when I wrote the proposal.

I've proposed attributes for the simple reason that HTML does not have templates, and HTML and PDF numbering should be the same.

**Comment by lotaris-xx** [comment]
This seems like a great idea. The trailing dots do bother me atm. I think the pair of attributes to specify the interior and trailing delimiter would be ideal.

**Comment by mojavelinux** [comment]
The more I think about this, the more I'm convinced it should be done via attributes because multiple converters will need to use it.

This change will need to be applied to core, but since Asciidoctor PDF is already overriding the method, we can change it here first, then follow-up in core.

**Comment by mojavelinux** [comment]
> sectnumappendseparator

I thin this should be sectnumsuffix.

**Comment by mojavelinux** [comment]
I wonder if we should shorten sectnumseparator and sectnumsuffix to just numseparator and numsuffix. The reason is, numbers are also used for blocks like figures (e.g., `Figure 1. Title`). So you're going to get the trailing dot there too. Or perhaps we shouldn't couple these things and have a different set of attributes for the caption label. wdyt?

**Comment by mojavelinux** [comment]
@janosvitok Thanks for responding to my call to architecture, btw. Very helpful.

**Comment by habakuk007** [comment]
I'm strongly interested in this feature because of [ISO 2145](https://en.wikipedia.org/wiki/ISO_2145) and [DIN 1421](https://de.wikipedia.org/wiki/DIN_1421) (explained in Wikipedia article [Gliederung](https://de.wikipedia.org/wiki/Gliederung#Dezimale_Gliederung) - [translation](https://translate.google.com/translate?hl=&sl=de&tl=en&u=https%3A%2F%2Fde.wikipedia.org%2Fwiki%2FGliederung%23Dezimale_Gliederung)).

And I would prefer to have different suffixes for [figures](https://github.com/asciidoctor/asciidoctor-pdf/issues/858#issuecomment-529310927) because I need them like here: `Tabelle 1: Einstellungen der seriellen Schnittstelle`.

**Comment by ullenboom** [comment]
Is there any update on this feature?
Also belongs to https://github.com/asciidoctor/asciidoctor/issues/3702

**Comment by mojavelinux** [comment]
If there is an update, I can assure you I will post it here. If you want it sooner, you can help by answering the open questions I have floated. If we can't answer those questions, we can't implement a change.

**Comment by mojavelinux** [comment]
I appreciate the feedback and I'll give it some thought. My initial reaction is that this is introducing too much complexity into the theme. I don't think that more options to configure is better. More options means more things users have to learn and makes the converter feel too complicated/overwhelming for most. We've had tremendous success with keeping things as simple as possible and letting the converter do the work. It's better for advanced functionality like this to allow the converter to be extended (or Asciidoctor in general) in order to influence the behavior. That would mean extracting the logic that sets up the numbering into an overriddable method. But that can be easier said than done.

**Comment by r0ckarong** [comment]
> More options means more things users have to learn and makes the converter feel too complicated/overwhelming for most. 

I can only speak for myself but I see it from the other side of the coin. I don't really learn often about something I don't need to configure, it just doesn't come across my attention. If the defaults are sensible, then I most likely will never touch them. But when I have some extrinsic motivation (like requirements or a styleguide) to adhere to and find out there is just no simple way to do achieve it it's demotivating to say the least. It's really very much easier to ignore something that you don't care about than to invent something that you desperately need; especially when you've got time constraints.

**Comment by mojavelinux** [comment]
I understand your perspective, but also understand that I support Asciidoctor close to 24/7, and I can tell you from experience that adding more options makes it harder to support. I have time constraints too, and the more complex the configuration is, the more pressure it puts on those constraints.

**Comment by bmj2023** [comment]
TEX has the contrib package secdot that turns on the trailing dot after the section number.  For simplicity the attribute could be :secdot!: to turn off the trailing dot in Asciidoctor.  
There are two other commands that allow for adding the trailing dot by level (sectiondot) and turning on the dot and modifying the format of the trailing dot (sectionpunct).  
A more complicated attribute could be :secdot: level, string.  

https://ctan.org/tex-archive/macros/latex/contrib/secdot

**Comment by bz007** [comment]
The idea of *separator and *suffix is good. Quite simple and configurable. I would suggest also add list numbering in same behaviour.
